From 1de13e85a369d51f76e8b6ebbb036f86bcf4365e Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Tue, 14 Sep 2021 10:19:07 +0300
Subject: [PATCH] [SW-54874] gaudi2: arc: load preboot to sram region

Currently, preboot image is dma'ed from spi to scratchpad sram
in secureboot flow. Since scratchpad sram is limited and is used
for other stuff in FW, this patch add changes to load and run
preboot at sram location.

Change-Id: I69670de6e5735664b89e437ccc352046df19e2c7
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index c729040..464ac07 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -204,21 +204,22 @@
 	image_size = (u16) spi_flash_readl(flash_offset);
 
 	/*
-	 * Copy preboot image from flash to Scratchpad SRAM.
+	 * Copy preboot image from flash to SRAM.
 	 * Note that the offset to be provided to perform DMA must be
 	 * in granuality of 4B. ie. In order to write at
-	 * 0x7FFBFE1000ULL, the address to be passed is 0x7FFBFE0400ULL
-	 * instead of 0x7FFBFE1000ULL (right shift offest 0x1000 by 2).
+	 * 0x1000FFFFFD004000ull, the address to be passed is
+	 * 0x1000FFFFFD001000ull instead of 0x1000FFFFFD004000ull
+	 * (right shift offset 0x4000 by 2).
 	 */
 	dma_copy_image(image_size, flash_offset,
-		       (HL_SP_SRAM_BASE + (SP_SRAM_PREBOOT_OFFSET >> 2)),
-		       SPI_DMA_DST_HBW);
+			(SDRAM_FULL_BASE + (PPBOOT_SIZE >> 2)),
+			SPI_DMA_DST_HBW);
 	if (dma_wait_finish() == DMA_CRC_ERROR)
 		return ARCBOOT_BRINGUP_ERROR;
 
 	image_size -= SECUREBOOT_IMG_HEAD_FOOTER;
 	if (secureboot_verify_image(image_size,
-			(SP_SRAM_BASE + SP_SRAM_PREBOOT_OFFSET + IMG_LEN_SIZE))
+			(SDRAM_BASE + PPBOOT_SIZE + IMG_LEN_SIZE))
 			== SECUREBOOT_VERIFY_FAILED)
 		return ARCBOOT_BRINGUP_ERROR;
 
diff --git a/arch/gaudi2/arc/common.h b/arch/gaudi2/arc/common.h
index 6635479..8aa141b 100644
--- a/arch/gaudi2/arc/common.h
+++ b/arch/gaudi2/arc/common.h
@@ -42,7 +42,7 @@
 #define PREBOOT_START_OFFSET	0x8
 #define PPBOOT_START_OFFSET	0xC
 #define PPBOOT_TEXT_BASE	(FW_SRAM_BASE + PPBOOT_START_OFFSET)
-#define PREBOOT_TEXT_BASE	(SP_SRAM_BASE + SP_SRAM_PREBOOT_OFFSET + PREBOOT_START_OFFSET)
+#define PREBOOT_TEXT_BASE	(SDRAM_BASE + PPBOOT_SIZE + PREBOOT_START_OFFSET)
 
 #define SP_SRAM_BOOT_INFO_OFFSET	0x1800
 #define BOOT_INFO_BASE		(SP_SRAM_BASE + SP_SRAM_BOOT_INFO_OFFSET)
